File: src/ui/sections/license_section.rs
//! The **License** status section: which tier this configuration runs at, and
//! the verified terms behind that answer.
//!
//! Read-only. Installing a licence happens in the setup wizard or the web
//! dashboard; this section exists so the terminal can answer "am I licensed,
//! and until when?" without either.
//!
//! Health maps the licence status to a semantic role rather than to the word:
//! a missing licence is the free tier working as intended, not a fault.

use crate::licensing::LicenseStatus;
use crate::ui::status_panel::{
    ActionMeta, ActionSet, SectionHealth, SectionId, StatusAction, StatusIcon, StatusSection,
    StatusSnapshot, TreeRow,
};

pub struct LicenseSection;

fn label_and_health(status: LicenseStatus) -> (&'static str, SectionHealth) {
    match status {
        LicenseStatus::Valid => ("Premium", SectionHealth::Green),
        LicenseStatus::Missing => ("Free", SectionHealth::Gray),
        LicenseStatus::Expired => ("Expired", SectionHealth::Yellow),
        LicenseStatus::NotYetValid => ("Not yet valid", SectionHealth::Yellow),
        LicenseStatus::Invalid => ("Invalid", SectionHealth::Red),
    }
}

fn timestamp(seconds: i64) -> String {
    chrono::DateTime::from_timestamp(seconds, 0)
        .map_or_else(|| "-".to_string(), |t| t.format("%Y-%m-%d").to_string())
}

impl StatusSection for LicenseSection {
    fn section_id(&self) -> SectionId {
        SectionId::License
    }

    fn label(&self) -> &'static str {
        "License"
    }

    fn prerequisites(&self) -> &[SectionId] {
        // Always answerable: the free tier is a valid answer.
        &[]
    }

    fn health(&self, snapshot: &StatusSnapshot) -> SectionHealth {
        label_and_health(snapshot.license.status).1
    }

    fn description(&self, snapshot: &StatusSnapshot) -> String {
        let (label, _) = label_and_health(snapshot.license.status);
        match &snapshot.license.terms {
            Some(terms) => format!("{label} · expires {}", timestamp(terms.exp)),
            None => format!("{label} · local execution included"),
        }
    }

    fn children(&self, snapshot: &StatusSnapshot) -> Vec<TreeRow> {
        let health = self.health(snapshot);
        let mut rows = vec![row(
            "license-configuration",
            "Configuration",
            &snapshot.license.profile_id.to_string(),
            StatusIcon::Key,
            health,
        )];

        if let Some(terms) = &snapshot.license.terms {
            rows.push(row(
                "license-subject",
                "Licensed to",
                &terms.sub,
                StatusIcon::Check,
                health,
            ));
            rows.push(row(
                "license-id",
                "License ID",
                &terms.jti,
                StatusIcon::File,
                health,
            ));
            rows.push(row(
                "license-validity",
                "Valid",
                &format!("{} to {}", timestamp(terms.nbf), timestamp(terms.exp)),
                StatusIcon::File,
                health,
            ));
        } else {
            rows.push(row(
                "license-free",
                "Included",
                "Multiple local agents and local containers",
                StatusIcon::Check,
                SectionHealth::Gray,
            ));
        }

        if let Some(url) = purchase_url(snapshot) {
            rows.push(TreeRow {
                section_id: SectionId::License,
                id: "license-purchase".to_string(),
                depth: 1,
                label: "Operator Premium".to_string(),
                description: url.clone(),
                icon: StatusIcon::Plug,
                brand_icon: None,
                is_header: false,
                actions: ActionSet {
                    primary: StatusAction::OpenUrl(url),
                    back: StatusAction::None,
                    special: StatusAction::None,
                    special_meta: None,
                    refresh: StatusAction::None,
                    refresh_meta: None,
                },
                health: SectionHealth::Gray,
            });
        }
        rows
    }
}

/// Only an https destination is ever offered as a link.
fn purchase_url(snapshot: &StatusSnapshot) -> Option<String> {
    snapshot
        .license
        .purchase_url
        .as_ref()
        .filter(|url| url.starts_with("https://"))
        .cloned()
}

fn row(
    id: &str,
    label: &str,
    description: &str,
    icon: StatusIcon,
    health: SectionHealth,
) -> TreeRow {
    TreeRow {
        section_id: SectionId::License,
        id: id.to_string(),
        depth: 1,
        label: label.to_string(),
        description: description.to_string(),
        icon,
        brand_icon: None,
        is_header: false,
        actions: ActionSet {
            primary: StatusAction::None,
            back: StatusAction::None,
            special: StatusAction::None,
            special_meta: None,
            refresh: StatusAction::None,
            refresh_meta: None,
        },
        health,
    }
}

#[allow(dead_code)]
fn unused(_: ActionMeta) {}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::config::Config;

    #[test]
    fn a_missing_licence_reads_as_free_not_broken() {
        let directory = tempfile::tempdir().unwrap();
        let mut config = Config::default();
        config.paths.state = directory.path().to_string_lossy().into_owned();
        let snapshot = StatusSnapshot::from_config(&config, vec![]);

        assert_eq!(LicenseSection.health(&snapshot), SectionHealth::Gray);
        assert!(LicenseSection.description(&snapshot).starts_with("Free"));
        assert!(LicenseSection.prerequisites().is_empty());
    }

    #[test]
    fn the_configuration_binding_is_always_shown() {
        let directory = tempfile::tempdir().unwrap();
        let mut config = Config::default();
        config.paths.state = directory.path().to_string_lossy().into_owned();
        let snapshot = StatusSnapshot::from_config(&config, vec![]);

        let rows = LicenseSection.children(&snapshot);
        assert!(rows.iter().any(|r| r.label == "Configuration"));
    }
}

File: vscode-extension/src/sections/license-section.ts
import * as vscode from "vscode";
import { StatusItem } from "../status-item";
import type { SectionContext, StatusSection } from "./types";
import type { SectionId, SectionHealth } from "../generated";
import type { LicenseResponse } from "../generated/LicenseResponse";
import { discoverApiUrl, OperatorApiClient } from "../api-client";

/**
 * License section - which tier this configuration runs at.
 *
 * Read-only, like the rest of the extension's status tree: installing a licence
 * happens in the hosted Operator UI, which the rows link out to. A missing
 * licence is the free tier working as intended, so it reads Gray rather than
 * as a fault.
 */
const STATUS_LABELS: Record<LicenseResponse["status"], string> = {
  missing: "Free",
  valid: "Premium",
  expired: "Expired",
  not_yet_valid: "Not yet valid",
  invalid: "Invalid",
};

const STATUS_HEALTH: Record<LicenseResponse["status"], SectionHealth> = {
  missing: "Gray",
  valid: "Green",
  expired: "Yellow",
  not_yet_valid: "Yellow",
  invalid: "Red",
};

const day = (seconds: bigint): string =>
  new Date(Number(seconds) * 1000).toISOString().slice(0, 10);

export class LicenseSection implements StatusSection {
  readonly sectionId: SectionId = "license";
  readonly prerequisites: SectionId[] = [];

  private license: LicenseResponse | null = null;

  health(): SectionHealth {
    return this.license ? STATUS_HEALTH[this.license.status] : "Gray";
  }

  async check(ctx: SectionContext): Promise<void> {
    try {
      const client = new OperatorApiClient(await discoverApiUrl(ctx.ticketsDir));
      this.license = await client.license();
    } catch {
      this.license = null;
    }
  }

  getTopLevelItem(_ctx: SectionContext): StatusItem {
    const label = this.license ? STATUS_LABELS[this.license.status] : "API required";
    return new StatusItem({
      label: "License",
      description: label,
      icon: "key",
      collapsibleState: this.license
        ? vscode.TreeItemCollapsibleState.Collapsed
        : vscode.TreeItemCollapsibleState.None,
      sectionId: this.sectionId,
      health: this.health(),
    });
  }

  getChildren(_ctx: SectionContext, _element?: StatusItem): StatusItem[] {
    if (!this.license) {
      return [];
    }
    const rows: StatusItem[] = [
      this.row("Configuration", this.license.profile_id, "symbol-namespace"),
    ];
    if (this.license.terms) {
      rows.push(
        this.row("Licensed to", this.license.terms.sub, "account"),
        this.row("License ID", this.license.terms.jti, "file"),
        this.row(
          "Valid",
          `${day(this.license.terms.nbf)} to ${day(this.license.terms.exp)}`,
          "calendar",
        ),
      );
    } else {
      rows.push(this.row("Included", "Multiple local agents and local containers", "check"));
    }
    return rows;
  }

  private row(label: string, description: string, icon: string): StatusItem {
    return new StatusItem({
      label,
      description,
      icon,
      collapsibleState: vscode.TreeItemCollapsibleState.None,
      sectionId: this.sectionId,
      health: this.health(),
      command: { command: "operator.openLicense", title: "Open License in Operator UI" },
    });
  }
}
